Htm advanced doc update

Zbysekz opened the pull request:
It was throwing errors when i tried to run examples, so 
doc update and mainly the problem with parsing numpy to string (i've started this question: https://stackoverflow.com/questions/62831299/python-numpy-to-string-weird-whitespaces/62832658?noredirect=1#comment111112331_62832658)
solved by specifying formatter.

Zbysekz commented:
can't add https://github.com/fcr user as reviewer, not sure why?

Zbysekz commented:
Should notify https://github.com/fcr

breznak commented on py/htm/advanced/README.md:
```diff
@@ -27,3 +27,7 @@ Connections
+
+Install python dependencies
+===========
+In order to run this python code, one must install requirements in htm/advanced/requirements.txt.
```
doc is always good :+1: 
It must be used somewhere, as we do run the advanced_tests (?)

Zbysekz replied:
The advanced_test don't need pillow & scikit-image, only advanced examples need this

breznak commented on py/htm/advanced/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+#See http://www.pip-installer.org/en/latest/requirements.html for details
+pillow>=7.2.0
+scipy>=1.5.1
```
nit: can you find some lower versions? For us as a lib, it's always good to keep reasonably old requirements. (but no big deal)

breznak commented:
there's a `extras` section in setup.py run as `pip install htm.core[examples]` which installs additional dependencies. I wonder if the "advanced" is in there as well?

fcr commented:
I noticed that @Zbysekz tried to modify the regions.extractList parser. Unfortunately, this parser is a quick and dirty (and fragile) work around to the compensate for the py_Region.execute command in the binding just converting arguments into default strings. Sometimes this conversion is not totally predictable. 
The parser could definitely be improved, but rather than just make a better work around, it would be much better if both ends used a more robust encoder/decoder to marshall the arguments .

dkeeney commented:
The execute command is a problem.  The call interface for the execute command is part of the Plug-in API.  All regions must implement it, C++ regions as well as Python.  That is the reason for the strings as arguments, because they are a common denominator. 

I am open to any wisardry that could allow arbitrary argument types and arbitrary return value type that would still work with C++.  Being able to call an arbitrary function would be nice too.

Perhaps using JSON would be an improvement.

Zbysekz commented on py/htm/advanced/frameworks/location/location_network_creation.py:
```diff
@@ -30,6 +30,8 @@
 from htm.advanced.frameworks.location.path_integration_union_narrowing import computeRatModuleParametersFromCellCount
 from htm.advanced.support.logging_decorator import LoggingDecorator
 
+np.set_printoptions(formatter={'float': '{: 0.3f}'.format})
```
@fcr firstly i tried modify the fragile parser, by adding more conditions for replacing characters, but then i figured out why it is behaving so weirdly... This statement `np.set_printoptions(formatter={'float': '{: 0.3f}'.format})` will set up proper formatting for numpy native conversion to string. See this [question](https://stackoverflow.com/questions/62831299/python-numpy-to-string-weird-whitespaces) that i have started. With this the example works without any of my further modifications.

@dkeeney thanks for explanation, i have concluded the same, that it has difficult background why it is like this. 
Anyway i am just digging around, just trying to run it on my machine, so no further changes now

fcr replied:
@Zbysekz,  the numpy print option is not relevant here since it is the C++ side that is generating the strings that have to be parsed in region. If it was the Python side generating the arguments there wouldn't even be a need to marshall and unmarshall the arguments.

Zbysekz replied:
> @Zbysekz, the numpy print option is not relevant here since it is the C++ side that is generating the strings that have to be parsed in region. If it was the Python side generating the arguments there wouldn't even be a need to marshall and unmarshall the arguments.

Maybe i have confused the problem by mentioning only the parser before...

My problem was with running l2l4l6_experiment.py in thing_classification example. The loading of "Things" is done on the python side and then [here](https://github.com/htm-community/htm.core/blob/907d60a44d3acf3348717b308cb1521e774297e4/py/htm/advanced/frameworks/location/location_network_creation.py#L516) it is passed further, then through some decorators and goes to [pyRegion.py](https://github.com/htm-community/htm.core/blob/907d60a44d3acf3348717b308cb1521e774297e4/bindings/py/packaging/src/htm/bindings/regions/PyRegion.py#L340) where the conversion from numpy array to string happens.
The error was raised when code in [advanced/regions/__init__.py](https://github.com/htm-community/htm.core/blob/master/py/htm/advanced/regions/__init__.py) file was trying to convert it to json

fcr replied:
This is exactly the problem I was talking about. 
The call to [pyRegion.py](https://github.com/htm-community/htm.core/blob/907d60a44d3acf3348717b308cb1521e774297e4/bindings/py/packaging/src/htm/bindings/regions/PyRegion.py#L340) then goes to [py_Engine.cpp](https://github.com/htm-community/htm.core/blob/907d60a44d3acf3348717b308cb1521e774297e4/bindings/py/cpp_src/bindings/engine/py_Engine.cpp#L302), the C++ arguments are converted to default strings which then eventually get to [advanced/regions/init.py](https://github.com/htm-community/htm.core/blob/master/py/htm/advanced/regions/__init__.py) where they need to be marshalled to Python objects.
The default strings are not consistent across C++ libraries. That is where the problem lies.

Zbysekz replied:
Yes that is unfortunate... but at least with this PR the numpy conversion would not make a problem anymore

Zbysekz commented:
> there's a `extras` section in setup.py run as `pip install htm.core[examples]` which installs additional dependencies. I wonder if the "advanced" is in there as well?

We have extras section in our setup.py but it seems wrong (maybe outdated format), i corrected it like it is in the [documentation](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies) and now i am able to install it correctly. Before i got always error: "There is no extras section in your setup.py". 

So i removed the requirements.txt that i have added into advanced.

Would be nice if somebody can confirm this as i am not so sure, but it seems good.

breznak commented on setup.py:
```diff
@@ -402,12 +402,12 @@ def configure(platform, build_type):
+    extras_require={'examples':['scikit-image>0.15.0',
+                    'sklearn',
+                    'matplotlib',
+                    'pillow',
```
thanks for correcting that. The old format used to work, so likely `pip` changed somewhere in the middle. 
Can you please check the advanced tests are run in the CI (I think so) - but that would require this method to be working even before..?

Zbysekz replied:
These are not needed for tests, only for examples, since these are plotting libs.
So i think it haven't have to work before. BTW i am not experienced with the CI...

Zbysekz replied:
sklearn is not for plots but it is used also in examples only 
![image](https://user-images.githubusercontent.com/23330229/87399130-33627900-c5b7-11ea-8d56-c934dd95171e.png)

Zbysekz commented:
@breznak ConnectionsPerformanceTest.testSP failed, but that's not connected with this PR right?

breznak commented:
Likely a random slow-down, restarting the tests. Sure, unrelated to the PR.

Zbysekz commented:
can be merged, if nobody has more questions
